### PR TITLE
Feat/order UI elements

### LINKED ISSUE #796: ui: order of elements in lists and tables
Currently things are ordered differently or randomly. We should specify how we order things in lists/tables etc... and make it consistent.

### PR DESCRIPTION
**Pull request by Thiemann96**
closes #796 

New input field to sort badges and insitutions. Additionally I added the green action bar to the badges catalog as well (same as the design suggestion for the issuer catalog page)

grafik

 - can be seen at the catalog pages for badges/issuers and the backpag page for badges

### CODE REVIEW

File: src/app/catalog/components/badge-catalog/badge-catalog.component.ts [hunk 1]
```diff
@@ -172,25 +201,24 @@ export class BadgeCatalogComponent extends BaseRoutableComponent implements OnIn
...
+		// Sortierfunktion
+		const sortFn = (a: any, b: any) => {
+			if (sortBy === 'name') {
+				return multiplier * a.name.localeCompare(b.name);
+			} else if (sortBy === 'date') {
+				const dateA = new Date(a.createdAt).getTime();
+				const dateB = new Date(b.createdAt).getTime();
+				return multiplier * (dateA - dateB);
+			}
+			return 0;
```
Comment (timber-they) [inline src/app/catalog/components/badge-catalog/badge-catalog.component.ts:217]: This case should be impossible, right? Maybe throw an exception or at least log something here then?

File: src/app/catalog/components/issuer-catalog/issuer-catalog.component.ts [hunk 2]
```diff
@@ -58,7 +52,15 @@ export class IssuerCatalogComponent extends BaseRoutableComponent implements OnI
 			value: 'andere',
 		},
 	];
-
+	labelDateAsc = this.translate.instant('General.dateAscending');
+	labelDateDesc = this.translate.instant('General.dateDescending');
+	sortOptions = [
```
Comment (timber-they) [inline src/app/catalog/components/issuer-catalog/issuer-catalog.component.ts:57]: This is a lot of duplicate code, that I'm sure could be extracted to a common helper or something similar

File: src/app/catalog/components/issuer-catalog/issuer-catalog.component.ts [hunk 3]
```diff
@@ -386,14 +391,24 @@ export class IssuerCatalogComponent extends BaseRoutableComponent implements OnI
...
-			this.issuerResultsByCategory.forEach((r) => r.issuers.sort((a, b) => a.name.localeCompare(b.name)));
-		} else {
-			this.issuerResults.sort((a, b) => b.name.localeCompare(a.name));
-			this.issuerResultsByCategory.forEach((r) => r.issuers.sort((a, b) => b.name.localeCompare(a.name)));
-		}
+	changeOrder(sortOption: string) {
+		const [sortBy, order] = sortOption.split('_') as ['name' | 'date', 'asc' | 'desc'];
+		const multiplier = order === 'asc' ? 1 : -1;
+
+		// Sortierfunktion
```
Comment (timber-they) [inline src/app/catalog/components/issuer-catalog/issuer-catalog.component.ts:398]: German comment

File: src/app/catalog/components/learningpath-catalog/learningpath-catalog.component.html [hunk 4]
```diff
@@ -14,34 +16,45 @@ <h1 hlmH1 class="tw-text-purple tw-font-extrabold">{{ 'LearningPath.catalogHeadl
...
-
+								placeholder="{{ 'LearningPath.searchLearningPaths' | translate }}"
+								[(ngModel)]="searchQuery"
+								hlmInput
+							/>
 						</div>
+
 						<oeb-checkbox
-							[text]="'Nach Institutionen gruppieren' "
+							[text]="'Nach Institutionen gruppieresssn'"
```
Comment (timber-they) [inline src/app/catalog/components/learningpath-catalog/learningpath-catalog.component.html:32]: I'm not sure that's an improvement

File: src/app/common/components/bg-badgecard.ts [hunk 5]
```diff
@@ -139,18 +152,13 @@ export class BgBadgecard {
...
 	@HostBinding('class') get completedClass(): string {
-		return this.completed
-			? 'tw-bg-[var(--color-green)]'
-			: 'tw-bg-white';
+		return this.completed ? 'tw-bg-[var(--color-green)]' : 'tw-bg-white';
 	}
 
-
-	showCompetencies = false
+	showCompetencies = false;
```
Comment (timber-they) [inline src/app/common/components/bg-badgecard.ts:161]: I'm just gonna assume that nothing except for formatting and similar changed in this file

File: src/app/recipient/components/recipient-earned-badge-list/recipient-earned-badge-list.component.ts [hunk 6]
```diff
@@ -84,18 +89,30 @@ export class RecipientEarnedBadgeListComponent extends BaseAuthenticatedRoutable
...
 	@ViewChild('badgesCounter') badgesCounter: CountUpDirective;
-	
+
 	activeTab = 'Badges';
 	private _badgesDisplay: BadgeDispay = 'grid';
 
+	labelDateAsc = this.translate.instant('General.dateAscending');
+	labelDateDesc = this.translate.instant('General.dateDescending');
+
+	sortOptions = [
```
Comment (timber-they) [inline src/app/recipient/components/recipient-earned-badge-list/recipient-earned-badge-list.component.ts:107]: Duplicate code

File: src/app/recipient/components/recipient-earned-badge-list/recipient-earned-badge-list.component.ts [hunk 7]
```diff
@@ -360,28 +380,43 @@ export class RecipientEarnedBadgeListComponent extends BaseAuthenticatedRoutable
...
-	// 			"Failed to load userProfile", error
-	// 		)
-	// 	);
-	// }
+	trackById(index: number, item: any): any {
+		return item.id; // oder eine andere eindeutige Eigenschaft
+	}
 
-	private groupCompetencies(badges) {
+	changeOrder(sortOption: string) {
```
Comment (timber-they) [inline src/app/recipient/components/recipient-earned-badge-list/recipient-earned-badge-list.component.ts:395]: ... and more duplicate code

File: src/assets/i18n/de.json [hunk 8]
```diff
@@ -59,7 +59,11 @@
 		"page": "Seite",
 		"learningPath": "Lernpfad",
 		"learningPaths": "Lernpfade",
-		"noLearningPaths": "Keine Lernpfade"
+		"noLearningPaths": "Keine Lernpfade",
```
Comment (timber-they) [inline src/assets/i18n/de.json:62]: Please also add English translations

File: src/app/catalog/components/badge-catalog/badge-catalog.component.ts [hunk 9]
```diff
@@ -39,6 +40,21 @@ export class BadgeCatalogComponent extends BaseRoutableComponent implements OnIn
...
+	sortOptions = [
+		{ value: 'name_asc', label: 'A-Z' },
+		{ value: 'name_desc', label: 'Z-A' },
+		{ value: 'date_asc', label: 'Älteste zuerst' },
+		{ value: 'date_desc', label: 'Neueste zuerst' },
+	];
+
+	sortControl = new FormControl('name_asc');
+
+	groupOptions = [
```
[the remark this answers is not in the document]
Comment (Thiemann96) [reply src/app/catalog/components/badge-catalog/badge-catalog.component.ts:52]: With adding a selection for sorting, I also adjusted the grouping selection to use the <oeb-select> component and and the search bar (see picture provided). Did not think this constitutes opening another card + issue

File: src/app/catalog/components/badge-catalog/badge-catalog.component.ts [hunk 10]
```diff
@@ -39,6 +40,21 @@ export class BadgeCatalogComponent extends BaseRoutableComponent implements OnIn
 	issuers: string[] = [];
 	selectedTag: string = null;
 
+	sortOptions = [
+		{ value: 'name_asc', label: 'A-Z' },
+		{ value: 'name_desc', label: 'Z-A' },
+		{ value: 'date_asc', label: 'Älteste zuerst' },
```
[the remark this answers is not in the document]
Comment (Thiemann96) [reply src/app/catalog/components/badge-catalog/badge-catalog.component.ts:46]: What do you mean exactly? Could you point me to place in the code where we do this? I edited the labels to use translations but I'm not sure thats what you mean?

File: src/app/catalog/components/badge-catalog/badge-catalog.component.html [hunk 11]
```diff
@@ -7,105 +7,154 @@
...
+				></oeb-background>
 			</div>
 		</div>
 
 		<app-badge-legend *ngIf="showLegend" (closed)="closeLegend()"></app-badge-legend>
 
 		<div class="tw-relative">
-
 			<section class="">
-				<div class="tw-flex tw-items-center oeb-standard-padding-bottom tw-flex-wrap">
```
[the remark this answers is not in the document]
Comment (Thiemann96) [reply src/app/catalog/components/badge-catalog/badge-catalog.component.html:25]: I know .. Umut told me to leave it on but we switched and will now do you code formating on the whole code on tuesday. Till then I leave it off

File: src/app/catalog/components/badge-catalog/badge-catalog.component.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Thiemann96) [reply src/app/catalog/components/badge-catalog/badge-catalog.component.ts:46]: I think I got it? I created a reusable <oeb-sort-select> component which reuses the code and applies proper translations (see `oeb-sort-select.component.ts` should also close comments about code duplication down below?

File: src/app/catalog/components/badge-catalog/badge-catalog.component.ts [hunk 12]
```diff
@@ -78,9 +94,15 @@ export class BadgeCatalogComponent extends BaseRoutableComponent implements OnIn
 		return this._groupBy;
 	}
 	set groupBy(val: string) {
+		console.log(val);
```
[the remark this answers is not in the document]
Comment (Thiemann96) [reply src/app/catalog/components/badge-catalog/badge-catalog.component.ts]: no was a leftover - is deleted

File: src/app/catalog/components/learningpath-catalog/learningpath-catalog.component.html (hunk 4 above)
Comment (Thiemann96) [reply src/app/catalog/components/learningpath-catalog/learningpath-catalog.component.html]: sorry for keeping that in my bad 🙃

File: src/app/recipient/components/recipient-earned-badge-list/recipient-earned-badge-list.component.ts (hunk 6 above)
Comment (Thiemann96) [reply src/app/recipient/components/recipient-earned-badge-list/recipient-earned-badge-list.component.ts]: see [my comment](https://github.com/mint-o-badges/badgr-ui/pull/879#discussion_r1863721117) above

**Comment by umut0** [comment]
- Things are not ordered by "newest" on first load, select field says A-Z
- options not visible in the dropdown 
image
- since the reordering takes quite some time i think we need to add a loading spinner or something like that. (Maybe we can improve load time too?)

**Comment by Thiemann96** [comment]
Have not found a way to reduce loading time of sorting. I'm not exactly sure why it takes so long but I guess it's because the 156 objetcs are very big (see below, wasnt even able to properly stringify it hence the picture) and it takes some time to iterate? 

I don't get why badges have to be in this format anyways, but I'm guessing there is a reason ... ? 

grafik

**Pull request merged**